### PR TITLE
chore: Extract bytecode from artifact files for backend integration test inputs

### PR DESCRIPTION
**Pull request by kevaundray**
# Description

This PR allows backends to not needed to parse the PreprocessedProgram. Instead they can parse the bytecode directly, this is inline with the fact that backends do not know about Noir, only ACIR.

## Problem\*

Resolves 

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by kevaundray** [comment]
> Is this just for the purposes of barretenberg testing? A better solution would be to extract the circuit from the json file inside `rebuild.sh`

I think this is also a less leaky abstraction because when backends run end to end tests, they will not want to parse the .json file because thats not a part of the ACVM interface

**Comment by kevaundray** [comment]
Note, we could also just make a change in the bb script which runs the tests, so it downloads the main.json and then parses the bytecode and deletes it for a file with just the bytecode.

Though then all backends would need to do that too.

Orthogonally, we'd also want to change the interface of bb.js if its expecting more than just the bytecode.

What are your thoughts?

**Comment by TomAFrench** [comment]
> Note, we could also just make a change in the bb script which runs the tests, so it downloads the main.json and then parses the bytecode and deletes it for a file with just the bytecode.
>
> Though then all backends would need to do that too.

My point is that `rebuild.sh` can do this step which removes the need for either Nargo or the backends' CI to do this.

**Comment by TomAFrench** [comment]
Unless we're expecting users to use this option in normal usage (in which case we need to discuss why) then it doesn't make sense for CI logic to leak into the binary we distribute to users.

**Comment by kevaundray** [comment]
> Unless we're expecting users to use this option in normal usage (in which case we need to discuss why) then it doesn't make sense for CI logic to leak into the binary we distribute to users.

It should be hidden to the user in this PR -- in either case, you could make an argument that the user may want to just compile the bytecode and witness, then run it with a different backend/binary, ie they only use Noir for compilation and then they call some other binary to consume it -- ie dynamic backends. 

Unless we are also expecting backends dynamic backends to take in the prepropressedProgram as input?

**Comment by kevaundray** [comment]
> > Note, we could also just make a change in the bb script which runs the tests, so it downloads the main.json and then parses the bytecode and deletes it for a file with just the bytecode.
> > Though then all backends would need to do that too.
> 
> My point is that `rebuild.sh` can do this step which removes the need for either Nargo or the backends' CI to do this.

Yep I understand this argument -- lets fallback onto this based on the conclusions we make to my previous question

**Comment by kevaundray** [comment]
Spoke with Tom and he made the good point that users will likely want to use the abi so they know how to encode the inputs. I think this is a good enough reason to switch to modifying the rebuild script and we can revisit adding bytecode-only flag when/if it ever comes up again as a viable workflow

**Review by TomAFrench — APPROVED** [review]

File: crates/nargo_cli/tests/execution_success/1327_concrete_in_generic/target/1327_concrete_in_generic.bytecode
Comment (phated) [inline crates/nargo_cli/tests/execution_success/1327_concrete_in_generic/target/1327_concrete_in_generic.bytecode:1]: This is very confusing and will fall out of sync without having nargo output split artifacts. We can move that higher up our priority list if it is needed now.

Comment (kevaundray) [reply crates/nargo_cli/tests/execution_success/1327_concrete_in_generic/target/1327_concrete_in_generic.bytecode:1]: Could you clarify what is confusing?

Comment (kevaundray) [reply crates/nargo_cli/tests/execution_success/1327_concrete_in_generic/target/1327_concrete_in_generic.bytecode:1]: I think the original PR allowed us to split the artifacts, but we changed to modifying the rebuild.sh script

Comment (phated) [reply crates/nargo_cli/tests/execution_success/1327_concrete_in_generic/target/1327_concrete_in_generic.bytecode:1]: Maybe I'm totally missing something but I don't see a `rebuild.sh` file being modified in this PR

File: crates/nargo_cli/tests/execution_success/rebuild.sh
Comment (TomAFrench) [inline crates/nargo_cli/tests/execution_success/rebuild.sh:1]: @phated

Comment (phated) [reply crates/nargo_cli/tests/execution_success/rebuild.sh:1]: It'd be nice if this file was in `crates/nargo_cli/tests` so it doesn't get lost when collapsing the massive directory tree in the github sidebar.

**Pull request merged**